### PR TITLE
Add a new HyphenTest unit test.

### PR DESCRIPTION
**Pull request by stoklund**
This test lives in its own .cpp file, but runs as part of the JITTest suite.
It trains and tests a small MLP network to hyphenate words.

### CODE REVIEW

File: tests/unittests/HyphenTest.cpp [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+  //
+  // Input: 6 x 27.
+  // Hidden: 16 fully connected, RELU output.
+  // Output: 1 fully connected, sigmoid output for logistic regression.
+  const size_t batchSize = numSamples;
+  ExecutionEngine EE(BackendKind::CPU);
+  auto &mod = EE.getModule();
+  Function *Infer = mod.createFunction("infer");
+
+  auto *InputVar =
```
Comment (nadavrot) [inline tests/unittests/HyphenTest.cpp]: The convention that we use is to start variable names with a lower case letter, unless it's an abbreviation (such as EE or M). Here we differ from the LLVM coding standards and use the FB coding standards.

Comment (stoklund) [reply tests/unittests/HyphenTest.cpp]: Thanks, will fix.

File: tests/unittests/HyphenTest.cpp [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                         VisibilityKind::Public, Variable::TrainKind::None);
+  auto *OutputVar =
+      mod.createVariable(ElemKind::FloatTy, {batchSize, 1}, "output",
+                         VisibilityKind::Public, Variable::TrainKind::None);
+
+  auto *HiddenFC = Infer->createFullyConnected("hidden_fc", InputVar, 16);
+  auto *Hidden = Infer->createRELU("hidden", HiddenFC);
+  auto *OutputFC = Infer->createFullyConnected("output_fc", Hidden, 1);
+  auto *Sig = Infer->createSigmoid("sig", OutputFC);
+  auto *Output = Infer->createRegression("regr", Sig, OutputVar);
```
Comment (artemrakhov) [inline tests/unittests/HyphenTest.cpp]: Regression in Glow means linear regression. The net will minimize Total Squared Error, while we really have classification task. Let's use `Softmax`

File: tests/unittests/HyphenTest.cpp [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+  for (size_t i = 0; i != numSamples; i++) {
+    float value = outHandle.at({i, 0});
+    if ((value > 0.5) == hyphens[i]) {
+      good++;
+    } else {
+      bad++;
+    }
+  }
+  EXPECT_GT(good, numSamples / 3);
+  EXPECT_LT(bad, 5);
```
Comment (artemrakhov) [inline tests/unittests/HyphenTest.cpp]: What are values of `good` and `bad` that you see? Can we expect 100% accuracy?

File: tests/unittests/HyphenTest.cpp [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    mapLetterWindow(words[i], i, inputHandle);
+    outputHandle.at({i, 0}) = hyphens[i];
+  }
+
+  // Now build the network:
+  //
+  // Input: 6 x 27.
+  // Hidden: 16 fully connected, RELU output.
+  // Output: 1 fully connected, sigmoid output for logistic regression.
+  const size_t batchSize = numSamples;
```
Comment (artemrakhov) [inline tests/unittests/HyphenTest.cpp]: If `batchSize` is smaller, we can expect the network to converge to better accuracy and run faster. Let's try 5 or 15.

**Comment by stoklund** [comment]
@nadavrot The training phase takes 800 ms with the JIT, so I don't think I want to slow down the unit tests by training under the interpreter. But I could train with the JIT and then run inference on other backends after transferring the weights.

File: tests/unittests/HyphenTest.cpp [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+  }
+
+  // This depends on the training data, of course.
+  const size_t numSamples = 566;
+  ASSERT_EQ(hyphens.size(), numSamples);
+  ASSERT_EQ(words.size(), numSamples);
+
+  // Randomly shuffle the training data.
+  // This is required for stochastic gradient descent training.
+  for (size_t i = 0; i != numSamples; i++) {
```
Comment (rdzhabarov) [inline tests/unittests/HyphenTest.cpp]: this is nitpicking, but this does not result in a random shuffle of the data. Something like reservoir sampling will be better (just a single change in the upper bound in nextRandInt).

File: tests/unittests/HyphenTest.cpp [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+  auto *HiddenFC = Infer->createFullyConnected("hidden_fc", InputVar, 16);
+  auto *Hidden = Infer->createRELU("hidden", HiddenFC);
+  auto *OutputFC = Infer->createFullyConnected("output_fc", Hidden, 1);
+  auto *Sig = Infer->createSigmoid("sig", OutputFC);
+  auto *Output = Infer->createRegression("regr", Sig, OutputVar);
+  auto *Result = Infer->createSave("result", Output);
+
+  // Generate a gradient network and use it for training.
+  Function *Train = glow::differentiate(Infer, EE.getConfig());
```
Comment (rdzhabarov) [inline tests/unittests/HyphenTest.cpp]: there is also batchsize on the EE config level which needs to be set.

File: tests/unittests/HyphenTest.cpp (hunk 2 above)
Comment (stoklund) [reply tests/unittests/HyphenTest.cpp]: Thanks. `Softmax` requires two outputs for a binary classification, right? Is there no way of getting a single yes/no answer with logistic regression from one output neuron?

File: tests/unittests/HyphenTest.cpp (hunk 3 above)
Comment (stoklund) [reply tests/unittests/HyphenTest.cpp]: I get `bad=2` out of 566 samples. It is possible that they are inconsistent samples where we require different outputs for the same input. I'll check. Without inconsistencies, this network should be able to overfit such a small dataset.

File: tests/unittests/HyphenTest.cpp (hunk 4 above)
Comment (stoklund) [reply tests/unittests/HyphenTest.cpp]: Yeah, that would be nice. I couldn't figure out how to run SGD with mini-batches using the `EE.runBatch()` function. Does it support mini-batches automatically, or should I call it with `iterations=1` for each mini-batch?

File: tests/unittests/HyphenTest.cpp (hunk 5 above)
Comment (rdzhabarov) [reply tests/unittests/HyphenTest.cpp]: Yes, that's what i thought

File: tests/unittests/HyphenTest.cpp [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+  // Convert the training data to word windows and labels.
+  vector<string> words;
+  vector<bool> hyphens;
+  for (auto *hword : TrainingData) {
+    dehyphenate(hword, words, hyphens);
+  }
+
+  // This depends on the training data, of course.
+  const size_t numSamples = 566;
+  ASSERT_EQ(hyphens.size(), numSamples);
```
[the remark this answers is not in the document]
Comment (stoklund) [reply tests/unittests/HyphenTest.cpp]: The following code is the random-shuffle loop `for (size_t i = 0; i != numSamples; i++)`, so it would be unsafe to continue if the two vectors don't have the required size. I think `ASSERT_EQ` is right here since it aborts immediately on failure.

File: tests/unittests/HyphenTest.cpp (hunk 2 above)
Comment (artemrakhov) [reply tests/unittests/HyphenTest.cpp]: Yes, it will require 2 neurons since we have 2 classes. Maybe change last FC to depth=2

File: tests/unittests/HyphenTest.cpp (hunk 4 above)
Comment (artemrakhov) [reply tests/unittests/HyphenTest.cpp]: SGD should work fine. Please take a look at `nonLinearClassifier` unit test. Maybe you didn't specify `EE_.getConfig().batchSize`, and it didn't converge because of doing too big steps?

File: tests/unittests/HyphenTest.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (rdzhabarov) [reply tests/unittests/HyphenTest.cpp]: EXPECT_EQ - should also fail the test immediately. Up to you, just have not found a lot of usages of asserts/ASSERT_EQ where regular gtest EXPECTs could be used.

[the remark this answers is not in the document]
Comment (stoklund) [reply tests/unittests/HyphenTest.cpp]: EXPECT_EQ will mark the test as failing, but it keeps running. This means you can have multiple failures from a single test run, which can be useful. See [the googletest docs explanation](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#assertions).
